fix: don't trash optional dependencies that fail due to incompatible platform

paulrutter opened the pull request:
This PR is a proposed solution for multiple reported issues regarding failed optional dependencies that are platform specific.
See #4828, #7961 and possibly related PR #8077.

It can be reproduced as follows:

- `mkdir test && cd test`
- `npm init -y`
- `npm i --verbose rollup`
- Read package-lock.json - observe 18 OS/CPU@rollup/* package variants are listed
- `rm package-lock.json`
- `npm i --verbose`
- Read package-lock.json - observe only 1 OS/CPU @rollup/* package variant is now listed for the current platform

With this PR, the `package-lock.json` will remain to contain the initial 18 rollup related dependencies, as any optional failure due to incompatible platform (`EBADPLATFORM`) will _not_ lead to trashing the node.

The side-effect of this is that _all_ platform specific dependencies will be present on _all_  platforms (just an empty directory where the platform doesn't match). But it does solve the issue at hand.

I would need help writing a proper unit test for this scenario though. With the steps below, i did manage to test my PR and see that it has the desired effect:

- `node . i --verbose rollup`
- Read package-lock.json - observe 18 OS/CPU@rollup/* package variants are listed
- `rm package-lock.json`
- `node . i --verbose`
- Read package-lock.json - observe 18 OS/CPU@rollup/* package variants are listed still

It will show this in the logging (on my Windows based machine):

```
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-win32-ia32-msvc
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-win32-arm64-msvc
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-linux-x64-musl
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-linux-x64-gnu
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-linux-s390x-gnu
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-linux-riscv64-gnu
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-linux-powerpc64le-gnu
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-linux-loongarch64-gnu
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-linux-arm64-musl
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-linux-arm64-gnu
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-linux-arm-musleabihf
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-linux-arm-gnueabihf
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-freebsd-x64
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-freebsd-arm64
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-darwin-x64
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-darwin-arm64
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-android-arm64
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\@rollup\rollup-android-arm-eabi
npm verbose reify skip trashing optional dependency due to platform mismatch C:\git\npm-cli\node_modules\fsevents
```

paulrutter committed:
fix for #4828: don't trash optional dependencies that fail to to an incompatible platform, to prevent issues when either the lockfile is removed

paulrutter committed:
fix for #4828: move check to before loop

paulrutter committed:
fix for #4828: logging improvement

paulrutter committed:
fix for #4828: apply the same fix to build-ideal-tree and log when pruning is skipped

paulrutter committed:
fix for #4828: logging improvement

paulrutter committed:
fix for #4828: simplify code

paulrutter commented on workspaces/arborist/lib/arborist/build-ideal-tree.js:
```diff
@@ -1489,6 +1489,13 @@ This is a one-time fix-up, please be patient...
         throw node.errors[0]
       }
 
+      // Optional dependencies that fail to install due to an incompatible platform
```
Not entirely sure if the change in build-ideal-tree is required, as this code doesn't seem to be ran in my test case.
Nonetheless, i added it here to prevent pruning optional dependencies based on platform mismatch.

paulrutter commented:
@wraithgar would you be able to take a look at this PR? Thanks 👍

paulrutter commented:
@owlstronaut thanks for running the test suite. So, in this case, would it make sense to adjust the tests? Seems like these are incorrect, if we want to retain the entries in the lock file.

About code coverage, where would i need to add or extend the tests for build-ideal-tree? See also https://github.com/npm/cli/pull/8127#discussion_r1967185187

owlstronaut commented:
> @owlstronaut thanks for running the test suite. So, in this case, would it make sense to adjust the tests? Seems like these are incorrect, if we want to retain the entries in the lock file.
> 
> About code coverage, where would i need to add or extend the tests for build-ideal-tree? See also [#8127 (comment)](https://github.com/npm/cli/pull/8127#discussion_r1967185187)

Sure! Yeah, in this case we would need to update the snapshot for the new expectations from reify, and for the `isolated-mode`, the 2 failing tests probably need a helper similar to `setupRequire` that will expect to find the folders, but find them empty.

`workspaces/arborist/test/arborist/build-ideal-tree.js` will need a test added for the code coverage.

I still have to leave your questions about the "right thing" to do open to @wraithgar, I've been on the team less than a month 😸

paulrutter commented:
Working through the tests and comments now, hope to push something this evening.

paulrutter committed:
fix for #4828:
- fix tests by adjusting snapshots for reify, now that it doesn't trash optional dependencies that fail due to a platform mismatch
- fix tests for isolated-mode by adjusting the assertions to not expect failure, but expect an empty directory though in case of a platform mismatch

paulrutter committed:
fix for #4828:
- Remove unneeded code in build-ideal-tree, as it currently doesn't seem to ever hit this code
- Without this, the original issue is still fixed

paulrutter commented:
@owlstronaut Could you run the test suite again? I fixed all the snapshot and failing tests.

Besides, i removed the code in build-ideal-tree, as it never seems to hit (in unit tests, but also not when building locally). The code in build-ideal-tree is not entirely clear to me, as i cannot figure out if `loadFailures` could ever contain a `EBADPLATFORM` error or not. In unit tests, it's always `E404` or `ETARGET`, never `EBADPLATFORM`. I tried adding a testcase for this, but still didn't get `EBADPLATFORM`. Hence i removed the code in build-ideal-tree.

With these changes, the original issue is still verified to be fixed, see the changed snapshots.
The remaining question is if this is the desired approach to this issue.

jakebailey commented on workspaces/arborist/test/isolated-mode.js:
```diff
@@ -552,9 +552,10 @@ tap.test('failing optional deps are not installed', async t => {
   await arborist.reify({ installStrategy: 'linked' })
 
-  t.notOk(setupRequire(dir)('which'), 'Failing optional deps should not be installed')
+  t.ok(setupRequire(dir)('which'), 'Failing optional deps should not be installed, but the empty directory is there')
```
Am I right to say that the new behavior would be to leave empty directories for all of the optional deps that did not succeed? Hopefully there's a way to avoid this, since that would leave a load of empty dirs for a lot of native packages, and potentially confuse code which is scanning for these dirs to know which one to load for the current platform...

paulrutter replied:
That is correct indeed. This is used to rebuild the proper package-json.lock file based on the `node_modules` directory.
I don't know enough of NPM to surface a better solution, but maybe others have ideas on how this can prevented?

Instead of the empty directory, it could also keep a list of failed optionals in a separate file or something, but that would require far more changes i would assume.

jakebailey replied:
That's really unfortunate; I thought that npm stored a file inside `node_modules` that described the layout? It seems like a bad idea to rely on an empty dir existing, and I really do think it'll cause something to break.

paulrutter replied:
Yes, i can imagine breakage if packages rely on directory scanning without verifying the contents. 

If this is not the desired approach, i would need guidance from the maintainers on what would be the best way forward.

owlstronaut replied:
Yep, we'll have @wraithgar weigh in

paulrutter replied:
For what it's worth, i checked out `bindings` which i use myself in https://github.com/blueconic/node-oom-heapdump, and that library handles this just fine. Haven't checked how nodejs itself handles it, as it's able to load native modules out of the box nowadays. 

https://github.com/TooTallNate/node-bindings/blob/c8033dcfc04c34397384e23f7399a30e6c13830d/bindings.js#L111

Also, the `rollup` package doesn't rely on directory scanning: https://github.com/rollup/rollup/blob/d3e79bd79e6b53b01ef7d6535d6ea9ad9be714dd/native.js#L77

`sharp` seems safe as well: https://github.com/lovell/sharp/blob/1533bf995acda779313fc178d2b9d46791349961/lib/sharp.js#L23

jakebailey replied:
The main ones to check out are rollup and esbuild (the bulk of the people in the issue having this problem), but there's also swc, oxc, dprint, who have this same problem too.

paulrutter replied:
`eslint` https://github.com/evanw/esbuild/blob/b914dd30294346aa15fcc04278f4b4b51b8b43b5/lib/npm/node-platform.ts#L150 is safe for the happy path, but the fallback for fixing the platform issue does use directory scanning.

`swc` is safe: https://github.com/swc-project/swc/blob/b65387ac570c4bbb3b776a69bda810be862b434e/packages/core/binding.js#L4

`oxc` is safe: https://github.com/oxc-project/oxc/blob/bff83c96a21d582775b65d97123d3a2ffabf30be/napi/parser/bindings.js#L4

`dprint` is safe: https://github.com/dprint/dprint/blob/00e8f5e9895147b20fe70a0e4e5437bd54d928e8/deployment/npm/install_api.js#L23

paulrutter commented:
To summarize the discussion above:

* with this PR @jakebailey is concerned about breaking libraries that rely on directory scanning for loading native modules
* the table below shows the common referenced libraries from #4828 and if these would be affected by this PR

| Library | Is affected by PR | git reference |
|--------|--------|--------|
| `bindings` | No | [git](https://github.com/TooTallNate/node-bindings/blob/c8033dcfc04c34397384e23f7399a30e6c13830d/bindings.js#L111) |
| `rollup` | No | [git](https://github.com/rollup/rollup/blob/d3e79bd79e6b53b01ef7d6535d6ea9ad9be714dd/native.js#L77) |
| `esbuild` | Possibly for the fallback behavior, happy path is safe | [git](https://github.com/evanw/esbuild/blob/b914dd30294346aa15fcc04278f4b4b51b8b43b5/lib/npm/node-platform.ts#L150) |
| `swc` | No | [git](https://github.com/swc-project/swc/blob/b65387ac570c4bbb3b776a69bda810be862b434e/packages/core/binding.js#L4) |
| `oxc` | No | [git](https://github.com/oxc-project/oxc/blob/bff83c96a21d582775b65d97123d3a2ffabf30be/napi/parser/bindings.js#L4) |
| `dprint` | No | [git](https://github.com/dprint/dprint/blob/00e8f5e9895147b20fe70a0e4e5437bd54d928e8/deployment/npm/install_api.js#L23) |
| `sharp` | No | [git](https://github.com/lovell/sharp/blob/1533bf995acda779313fc178d2b9d46791349961/lib/sharp.js#L23) |

jakebailey commented:
I assume you meant esbuild in the table above; the other bit is just the "cleanliness" of having dirs for an unbounded number of platforms and the annoyance of having to find the nonempty one when browsing the tree.

jsolly commented:
Hi all. I am having the issue outlined in[ this closed ticket](https://github.com/npm/cli/issues/8106) which was closed as a dup.

It sounds like there's a fix that's coming soon? This is the workaround I am using in GitHub Actions with the new arm architecture runner, `runs-on: ubuntu-24.04-arm`

```
      - name: Install dependencies
        run: npm install
      - name: WORKAROUND (Install biome manually)
        run: npm install --no-save @biomejs/cli-linux-arm64
```

paulrutter committed:
Merge branch 'npm:latest' into maintenance/fix-4828-dont-thrash-optional-dependencies-due-to-incompatible-platform

paulrutter commented:
> Hi all. I am having the issue outlined in[ this closed ticket](https://github.com/npm/cli/issues/8106) which was closed as a dup.
> 
> It sounds like there's a fix that's coming soon? This is the workaround I am using in GitHub Actions with the new arm architecture runner, `runs-on: ubuntu-24.04-arm`
> 
> ```
>       - name: Install dependencies
>         run: npm install
>       - name: WORKAROUND (Install biome manually)
>         run: npm install --no-save @biomejs/cli-linux-arm64
> ```

Let's try to keep this PR clean, as the related issue #4828 already contains a lot of similar reports.
Thanks!

restjohn commented:
Could this also fix https://github.com/npm/cli/issues/7622? 🙏

paulrutter commented:
> Could this also fix #7622? 🙏

I doubt it will, as this PR won't change anything to the lockfile, except making sure it's consistent across different platforms.

moiz-frost commented:
Meanwhile what can I do to for a workaround please?

paulrutter commented:
@owlstronaut @wraithgar. Sorry for asking again, but is there any timeline on getting feedback on this PR? Thanks!

owlstronaut commented:
@paulrutter sorry for the delay, I was on vacation last week!

After some discussion with @wraithgar, we don't think this is quite what we want to do. It is clever, but feels like a bit of a workaround as it is really using side-effects to get the desired outcome.

The issue stems from the fact that, when the lockfile is missing, the build-ideal-tree first builds from what already exists in the node_modules, and attempts to change it as little as possible https://github.com/npm/cli/blob/latest/workspaces/arborist/lib/arborist/build-ideal-tree.js#L291-L312

To have subsequent package‑lock generations exactly match the original, you need to "reintroduce" into the ideal tree any optional dependencies that were recorded in the original lockfile, even if they weren't installed on the current platform. Rather than building the ideal tree solely from what's physically present in node_modules (which is why a platform‑mismatched dependency gets dropped), you'd modify the tree‐building process so that it merges in the missing optional dependency metadata.

What probably needs to happen is in [here](https://github.com/npm/cli/blob/latest/workspaces/arborist/lib/arborist/build-ideal-tree.js#L314-L341). Traverse to find any optional dependencies that are not in the tree. When one is found, a new Node is created and added to the ideal tree's inventory. This way, when you regenerate the package‑lock.json from the ideal tree, it will include both the installed package and the optional dependency that wasn't physically reified.

paulrutter commented:
Thanks for the elaborate feedback @owlstronaut. It makes sense that this is not the desired approach, i kinda thought so already. But since it did fix the issue at hand, it was a good way of getting this issue some attention. 

Not sure if i have the knowledge to pull of the suggestions made, but given all the related reports, shouldn't this be something that should be handled by the maintainers?
Happy to assist where i can, just not sure if i know enough from the npm ecosystem to get this through.

owlstronaut commented:
@paulrutter Yes, totally understand the desire to just get something working. Ideally this is something we, the maintainers, should handle - we just have limited resources to fund all of the priorities we face.

paulrutter commented:
Understood👍🏻i contribute to other open source projects, so i recognize the challenge.
Let's see how we can move forward with this then; i will think about the outlined desired solution a bit further as well.
When i'm feeling strong, i might do another attempt. Otherwise we need to wait until someone else can do the work required.

zkat commented:
@wraithgar sorry, I got pulled in because typescript stuff but:

How would you feel about keeping "pruned" optional deps in the physical tree (which I assume is what gets written to `node_modules/.package-lock.json`, and just flagging them with a bool or something instead of removing them from the tree entirely? You could get identical install behavior that way, but when the `package-lock.json` gets removed, you could still just grab `node_modules/.package-lock.json` and rebuild the ideal tree from that directly?

Arborist is (mostly) after my time, but I can't imagine this being a very big change to implement, and I don't mind sending a patch myself in the next couple of days if y'all consent to it. And hey, I'm on the payroll for now, so :)

zkat commented:
Hi. Just dropping a note that I'm trying to tackle this over at https://github.com/npm/cli/pull/8177 and I'm pretty confident in the approach but I don't be able to test in earnest until Monday. Some details aren't in place yet

paulrutter commented:
Closing this one in favor of #8177
